core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableNestedLoopJoin.java

```diff
@@ -0,0 +1,245 @@
+        left,
+        right,
+        condition,
+        variablesSet,
+        requiredColumns,
+        joinType);
+  }
+
+  @Override public RelOptCost computeSelfCost(RelOptPlanner planner,
+      RelMetadataQuery mq) {
+    double rowCount = mq.getRowCount(this);
+
+    // Right-hand input is the "build", and hopefully small, input.
+    final double rightRowCount = right.estimateRowCount(mq);
+    final double leftRowCount = left.estimateRowCount(mq);
+
+    if (isNonCorrelateSemiJoin()) {
+      if (Double.isInfinite(leftRowCount)) {
+        rowCount = leftRowCount;
+      } else {
+        rowCount += Util.nLogN(leftRowCount);
+      }
+      if (Double.isInfinite(rightRowCount)) {
+        rowCount = rightRowCount;
+      } else {
+        rowCount += rightRowCount;
+      }
+      return planner.getCostFactory().makeCost(rowCount, 0, 0).multiplyBy(.01d);
+    } else {
+      if (Double.isInfinite(leftRowCount) || Double.isInfinite(rightRowCount)) {
+        return planner.getCostFactory().makeInfiniteCost();
+      }
+
+      Double restartCount = mq.getRowCount(getLeft());
+      // RelMetadataQuery.getCumulativeCost(getRight()); does not work for
+      // RelSubset, so we ask planner to cost-estimate right relation
+      RelOptCost rightCost = planner.getCost(getRight(), mq);
+      RelOptCost rescanCost =
+          rightCost.multiplyBy(Math.max(1.0, restartCount - 1));
+
+      return planner.getCostFactory().makeCost(
+          rowCount /* generate results */ + leftRowCount /* scan left results */,
+          0, 0).plus(rescanCost);
+    }
+  }
+
+  @Override public Join copy(RelTraitSet traitSet, RexNode conditionExpr, RelNode left,
+      RelNode right, JoinRelType joinType, boolean semiJoinDone) {
+    return new EnumerableNestedLoopJoin(getCluster(), traitSet, left, right,
+        conditionExpr, variablesSet, requiredColumns, joinType);
+  }
+
+  @Override public RelWriter explainTerms(RelWriter pw) {
+    // This is a normal semi-join.
+    final RelWriter writer = pw.input("left", left)
+        .input("right", right);
+    if (isNonCorrelateSemiJoin()) {
+      return writer
+          .item("condition", condition)
+          .item("joinType", joinType.lowerName)
+          .itemIf(
+              "systemFields",
+              getSystemFieldList(),
+              !getSystemFieldList().isEmpty());
+    } else {
+      // This is a correlate semi-join.
+      return writer
+          .item("correlation", getVariablesSet().iterator().next())
+          .item("joinType", joinType.lowerName)
+          .item("requiredColumns", requiredColumns.toString());
+    }
+  }
+
+  public Result implement(EnumerableRelImplementor implementor,
+      Prefer pref) {
+    if (isNonCorrelateSemiJoin()) {
+      assert joinInfo.isEqui();
+      return implementSemiJoin(implementor, pref);
+    }
+    return implementCorrelate(implementor, pref);
+  }
+
+  private Result implementSemiJoin(EnumerableRelImplementor implementor,
```

Comment:
I think EnumerableNestedLoopJoin should be just a renaming of EnumerableCorrelate.
IMHO it does not make sense to include here this `implementSemiJoin` method, which actually uses `BuiltInMethod.SEMI_JOIN.method ` i.e. a "hash-semijoin", so we are mixing nested loop and hash. 
Maybe this `implementSemiJoin` block could be integrated inside EnumerableJoin (now renamed as EnumerableHashJoin): the `EnumerableHashJoin#implement` method could do something like:
```
if (type = SEMI)
  use BuiltInMethod.SEMI_JOIN.method
else
  use BuiltInMethod.JOIN.method, as usual
```

BTW, minor detail: maybe we should rename `BuiltInMethod.JOIN` and `BuiltInMethod.SEMI_JOIN` as `BuiltInMethod.HASH_JOIN` / `BuiltInMethod.HASH_SEMI_JOIN`

Reply:
I agree, will make modifications:
1. move `implementSemiJoin` to EnumerableHashJoin
2. rename `BuiltInMethod.JOIN` to `BuiltInMethod.HASH_JOIN`, `BuiltInMethod.SEMI_JOIN` to `BuiltInMethod.HASH_SEMI_JOIN`.

Reply:
I'm sorry, but I think we I said in my previous comment regarding BuiltInMethod.SEMI_JOIN implementation was not 100% true. I think its actual implementation depends on the innerLookup contains method, which may be backed by a hash set... or not; so maybe the renaming of BuiltInMethod.SEMI_JOIN as BuiltInMethod.HASH_SEMI_JOIN is not a good idea.

Reply:
Yeah, let's just keep it as BuiltInMethod.SEMI_JOIN